nautobot/ipam/utils/migrations.py

```diff
@@ -0,0 +1,379 @@
+
+    Args:
+        cidr (str): IPv4/IPv6 CIDR string
+        namespace (Namespace): Namespace instance
+        max_prefix_length (int): Maximum prefix length depth for closest parent lookup
+    """
+    # Validate that it's a real CIDR
+    cidr = validate_cidr(apps, cidr)
+    broadcast = str(cidr.broadcast or cidr.ip)
+    ip_version = cidr.version
+
+    try:
+        max_prefix_length = int(max_prefix_length)
+    except ValueError:
+        raise ValidationError({"max_prefix_length": f"Invalid prefix_length: {max_prefix_length}."})
+
+    # Walk the supernets backwrds from smallest to largest prefix.
+    try:
+        supernets = cidr.supernet(prefixlen=max_prefix_length)
+    except ValueError as err:
+        raise ValidationError({"max_prefix_length": str(err)})
+    else:
+        supernets.reverse()
+
+    # Enumerate all unique networks and prefixes
+    networks = {str(s.network) for s in supernets}
+    del supernets  # Free the memory because it could be quite large.
+
+    # Prepare the queryset filter
+    lookup_kwargs = {
+        "network__in": networks,
+        # This might be flawed if an IPAddress has a prefix_length that excludes it from a
+        # parent that should otherwise contain it.
+        "prefix_length__lte": cidr.prefixlen,
+        "ip_version": ip_version,
+        "broadcast__gte": broadcast,
+        "namespace": namespace,
+    }
+
+    # Search for possible ancestors by network/prefix, returning them in reverse order, so that
+    # we can choose the first one.
+    Prefix = apps.get_model("ipam", "Prefix")
+    possible_ancestors = Prefix.objects.filter(**lookup_kwargs).order_by("-prefix_length")
+
+    # If we've got any matches, the first one is our closest parent.
+    try:
+        return possible_ancestors[0]
+    except IndexError:
+        raise Prefix.DoesNotExist(f"Could not determine parent Prefix for {cidr}")
+
+
+def generate_parent_prefix(apps, address):
+    """For a given `address`, generate a containing parent network address."""
+    cidr = validate_cidr(apps, address)
+    return cidr.cidr
+
+
+def process_vrfs(apps, schema_editor):
+    VRF = apps.get_model("ipam", "VRF")
+    VRFDeviceAssignment = apps.get_model("ipam", "VRFDeviceAssignment")
+
+    vrfs = VRF.objects.all().order_by("name", "rd")
+    unique_vrfs = vrfs.filter(enforce_unique=True)
+    # These will all be in the Global NS
+    non_unique_vrfs = vrfs.exclude(enforce_unique=True)
+
+    print("\n", end="")
+
+    # Case 0: VRFs with enforce_unique move to their own Namespace.
+    for vrf in unique_vrfs:
+        print(f">>> Processing migration for VRF {vrf.name!r}, Namespace {vrf.namespace.name!r}")
+        vrf.namespace = create_vrf_namespace(apps, vrf)
+        vrf.save()
+        vrf.prefixes.update(namespace=vrf.namespace)
+        print(f"    VRF {vrf.name!r} migrated to Namespace {vrf.namespace.name!r}")
+
+    # Case 00: Unique fields vs. Namespaces (name/rd) move to a Cleanup Namespace.
```

Comment:
Case 1?

Reply:
Was mirroring the cases from the wiki page. Case 1 requires no changes. See: https://github.com/nautobot/nautobot/wiki/2.0:-IPAM-Overhaul-Data-Migration-Plan#need-to-determine-which-vrfs-become-namespaces